src/KPM.jl

```diff
@@ -7,55 +7,76 @@ struct MomentaKPM{T}
+`μ_n = Tr[A T_n(h)] ≈ ∑ₐ⟨a|A T_n(h)|a⟩/N` over `N = randomkets` normalized random `|a⟩`. 
+Furthermore, the trace over a specific set of kets can also be computed; in this case
+`ket::AbstractMatrix` must be a matrix where the columns are the kets involved in the calculation.
 
-The order of the Chebyshev expansion is `order`. For the global density of states the trace
-is estimated stochastically using a number `randomkets` of random vectors. The
-`bandbrange = (ϵmin, ϵmax)` should completely encompass the full bandwidth of `hamiltonian`.
-If `missing` it is computed automatically.
+The order of the Chebyshev expansion is `order`. The `bandbrange = (ϵmin, ϵmax)` should completely encompass 
+the full bandwidth of `hamiltonian`. If `missing` it is computed automatically using `ArnoldiMethods` (must be loaded).
 
 # Example
 ```
 julia> h = LatticePresets.cubic() |> hamiltonian(hopping(1)) |> unitcell(region = RegionPresets.sphere(10));
 
-julia> Elsa.MomentaKPM(bloch(h), bandrange = (-6,6))
+julia> momentaKPM(bloch(h), bandrange = (-6,6))
 Elsa.MomentaKPM{Float64}([0.9594929736144973, -0.005881595972403821, -0.4933354572913581, 0.00359537502632597, 0.09759451291347333, -0.0008081453185250322, -0.00896262538765363, 0.00048205637037715177, -0.0003705198310034668, 9.64901673962623e-20, 9.110915988898614e-18], (0.0, 6.030150753768845))
 ```
 """
-momentaKPM(h::AbstractMatrix; ket = missing, kw...) = _momentaKPM(h, ket; kw...)
+momentaKPM(h::AbstractMatrix; ket = missing, obs = missing, kw...) = _momentaKPM(h, ket, obs; kw...)
 
-function _momentaKPM(h::AbstractMatrix{Tv}, ket::AbstractVector{T}; order = 10, bandrange = missing, kw...) where {T,Tv}
+function _momentaKPM(h::AbstractMatrix{Tv}, ket::AbstractVector{T}, obs; order = 10, bandrange = missing, kw...) where {T,Tv}
     μlist = zeros(real(promote_type(T, Tv)), order + 1)
     bandbracket = _bandbracket(h, bandrange)
     ket0 = normalize(ket)
-    _addmomenta!(μlist, ket0, h, bandbracket)
+    _addmomenta!(μlist, ket0, h, bandbracket, obs)
     return MomentaKPM(jackson!(μlist), bandbracket)
 end
 
-function _momentaKPM(h::AbstractMatrix{Tv}, ket::Missing; randomkets = 1, order = 10, bandrange = missing, kw...) where {Tv}
+function _momentaKPM(h::AbstractMatrix{Tv}, ket::AbstractMatrix{T}, obs; order = 10, bandrange = missing, kw...) where {T,Tv}
+    v = zeros(Tv, size(h,1))
+    μlist = zeros(real(Tv), order + 1)
+    bandbracket = _bandbracket(h, bandrange)
+    numkets = Int64(sum(nonzeros(ket)))
+    normfactor = 0.0
+    [_addmomenta!(μlist, Array{Tv}(ket[:,n]), h, bandbracket, obs) for n in 1:numkets]
+    if isa(obs, AbstractMatrix) == true
+        [normfactor += nonzeroweight(ket[:,n], obs) for n in 1:numkets]
+        μlist ./= normfactor
+    else  nothing end
+    return MomentaKPM(jackson!(μlist), bandbracket)
+end
+
+#Comment 1: the two previous methods of _momentaKPM could have been merged into one.
```

Comment:
Yeah, I think we can simplify this quite a bit. I'll propose a concrete change in a bit.